packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts

```diff
@@ -496,4 +652,167 @@ export default class UserStorageController extends BaseController<
       state.isProfileSyncingUpdateLoading = isProfileSyncingUpdateLoading;
     });
   }
+
+  /**
+   * Syncs the internal accounts list with the user storage accounts list.
+   * This method is used to make sure that the internal accounts list is up-to-date with the user storage accounts list and vice-versa.
+   * It will add new accounts to the internal accounts list, update/merge conflicting names and re-upload the results in some cases to the user storage.
+   */
+  async syncInternalAccountsWithUserStorage(): Promise<void> {
+    if (!this.#accounts.isAccountSyncingEnabled) {
+      return;
+    }
+
+    try {
+      this.#assertProfileSyncingEnabled();
+
+      await this.#accounts.setIsUserStorageAccountSyncingInProgress(true);
+
+      const userStorageAccountsList =
+        await this.#accounts.getUserStorageAccountsList();
+
+      if (!userStorageAccountsList || !userStorageAccountsList.length) {
+        await this.#accounts.saveInternalAccountsListToUserStorage();
+        await this.#accounts.setIsUserStorageAccountSyncingInProgress(false);
+        return;
+      }
+
+      // Compare internal accounts list with user storage accounts list
+      // First step: compare lengths
+      let internalAccountsList = await this.#accounts.getInternalAccountsList();
+
+      if (!internalAccountsList || !internalAccountsList.length) {
+        throw new Error(`Failed to get internal accounts list`);
+      }
+
+      const hasMoreInternalAccountsThanUserStorageAccounts =
+        internalAccountsList.length > userStorageAccountsList.length;
+
+      // We don't want to remove existing accounts for a user
+      // so we only add new accounts if the user has more accounts than the internal accounts list
+      if (!hasMoreInternalAccountsThanUserStorageAccounts) {
+        const numberOfAccountsToAdd =
+          userStorageAccountsList.length - internalAccountsList.length;
+
+        // Create new accounts to match the user storage accounts list
+        const addNewAccountsPromises = Array.from({
+          length: numberOfAccountsToAdd,
+        }).map(async () => {
+          await this.messagingSystem.call('KeyringController:addNewAccount');
+        });
+
+        await Promise.all(addNewAccountsPromises);
+      }
+
+      // Second step: compare account names
+      // Get the internal accounts list again since new accounts might have been added in the previous step
+      internalAccountsList = await this.#accounts.getInternalAccountsList();
+
+      for await (const internalAccount of internalAccountsList) {
+        const userStorageAccount = userStorageAccountsList.find(
+          (account) => account.a === internalAccount.address,
+        );
+
+        if (!userStorageAccount) {
+          await this.#accounts.saveInternalAccountToUserStorage(
```

Comment:
Future/perf - we can refactor this once we support batch.

We'll be mindful if there are users with many accounts.